### PR TITLE
Post Pallet for Storing "Remarks" On-Chain

### PR DESCRIPTION
**Pull request by shawntabrizi**
This is a simple, but quite needed pallet allowing users to store some arbitrary data on-chain for a deposit.

This is most similar to `remark`, where a user can post some arbitrary comment, but unlike remarks the data in the `pallet-post` gets stored on chain. This will allow 3rd party developers to bootstrap a number of scenarios on top of Substrate based networks without needing to create custom pallets.

In order to avoid post conflicts, all posts are gated by two keys: AccountId and Topic. There are two kinds of posts depending on the ordering of these keys:

* Blogs: Account -> Topic -> Post
* Threads: Topic -> Account -> Post

The idea of blogs is that a user may want to make a post on a number of different topics, just like a personal blog. In this kind of post format, you can easily query all the posts from a user by querying for all the blogs under an Account, which will give you a list of topics, and finally the underlying post.

The idea of threads is that many users may want to make a post on a shared topic, just like a forum thread. In this kind of post format, you can easily query all of the topics that are open for discussion, and then the accounts that have commented on that topic and made a post. Unfortunately there isnt directly a way to make multiple posts on the same topic, but in combination with blogs, such a design could probably be hacked together. Otherwise, we may want to consider a triple map, which is `Topic -> Account -> Index -> Post`.

Users will pay a deposit base on the number of bytes used in their topic + post. To return that deposit, users simply need to delete their post.

Posts include a block_number so they may be easily sorted and presented to the outside world.

## Scenarios

* We want to make a poll where on-chain users will provide feedback about some general topic. Off-chain we will define a certain "topic" for discussion, and a format for how users should make a post (for example sending `0x00` for nay and `0x01` for aye. Then users can make a post under that thread for the duration of the polling time. At the end of the polling time, a snapshot of the chain and the current votes can be tallied, with any filtering applied off-chain (for example a filter for only accounts with identities or a weighting factor based on number of tokens held at the end of the poll time. As long as these criteria are made public, anyone else could trustlessly verify the results.

* We want to create a "profile page" for all validators to make comments about their validator setup and why users should vote for them. Here we make another topic under a thread where anyone can make a post using some defined format. Then a service like Polkadot JS could query all the pages under this thread, filtering for users who are in the staking pool, and show this information for each user.

* I may want to make a series of personal posts about things happening on the network. I would add a new topic under my blogs for each of my comments. I could then host a page which queries and presents all of these posts under my account. I could also use this mechanism as a simple way to index into my posts in other threads. For example, I can make a post in a thread under a certain topic, and in my blog place a simple indicator like `0x01` under the same topic so show that if you look at the thread, you would find my post. (i.e. `account -> topic -> "0x01"` could imply that `topic -> account -> post exists`

* Users could post on chain details about some democracy proposal, or whether they support such a proposal using posts, in addition to depending on 3rd party services.

Fixes: https://github.com/paritytech/substrate/issues/7109

### CODE REVIEW

File: frame/post/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+		fn post(origin, post_type: PostType, topic: Topic, post: Vec<u8>) {
+			let poster = ensure_signed(origin)?;
+
+			let topic_len = topic.len();
+			ensure!(topic_len <= T::MaxTopicLength::get(), Error::<T>::TopicLength);
+			let post_len = post.len();
+			ensure!(post_len <= T::MaxPostLength::get(), Error::<T>::PostLength);
+
+			// Deposit is at least MinDeposit, or grows with topic and post length.
+			let deposit = T::ByteDeposit::get().saturating_mul(
```
Comment (xlc) [inline frame/post/src/lib.rs:120]: I think `ByteDeposit * (topic_len + post_len) + BaseDeposit` makes more sense over  `max(ByteDeposit * (topic_len + post_len), MinDeposit)`?

So the formula is more similar to other bytes deposit like treasury tip reason

File: frame/post/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+		type Error = Error<T>;
+
+		fn deposit_event() = default;
+
+		#[weight = T::WeightInfo::post()]
+		fn post(origin, post_type: PostType, topic: Topic, post: Vec<u8>) {
+			let poster = ensure_signed(origin)?;
+
+			let topic_len = topic.len();
+			ensure!(topic_len <= T::MaxTopicLength::get(), Error::<T>::TopicLength);
```
Comment (xlc) [inline frame/post/src/lib.rs:115]: I hope there is an easy way to have those checks in `validate` stage so bad tx will get rejected straightaway by tx pool.

It is possible to achieve this with SignedExtension but that requires too much boilerplate code

Comment (shawntabrizi) [reply frame/post/src/lib.rs:115]: why does it matter that it gets rejected in validate stage?
If it does not pass this test, it just looks like a remark?

Comment (shawntabrizi) [reply frame/post/src/lib.rs:115]: Though, i agree, would be cool to introduce some magic to make validation processes (like even ensure_signed) appear somewhere else in an easy to manage way.

Comment (xlc) [reply frame/post/src/lib.rs:115]: There is not need to include transactions into block that we know for sure that will fail. Saves people some tx fee.

Comment (shawntabrizi) [reply frame/post/src/lib.rs:115]: so this is a more general comment, because the same problem exists for nearly all our extrinsics.

Comment (xlc) [reply frame/post/src/lib.rs:115]: Yes

**Comment by marcio-diaz** [comment]
Very much needed pallet indeed. Fwiw, looks good to me. 

Perhaps another way to think about these "social structures" is in terms of trees, with roots and parent links. Then a "topic" may be just a post with a special null root (zero?). I'm thinking something like this:

account -> [postId]
parentId -> [postId]
postId -> post { parentId, account, content, deposit }
nextId

with deposit according to content. But I'm aware that is not efficient as your version, maybe you can comment on the cons.

File: frame/post/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+	type MaxPostLength: Get<usize>;
+	/// Weight information for extrinsics in this pallet.
+	type WeightInfo: WeightInfo;
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebug, Default)]
+pub struct Post<Balance, BlockNumber> {
+	pub deposit: Balance,
+	pub block_number: BlockNumber,
+	pub post: Vec<u8>,
```
Comment (marcio-diaz) [inline frame/post/src/lib.rs:61]: or
```suggestion
	pub content: Vec<u8>,
```

**Comment by shawntabrizi** [comment]
@marcio-diaz the main goal here should be to make it easy to query all posts under a topic.

The main scenarios this hopes to solve are related to having users all post some details about some topic.

For example:

* Validators posting why you should nominate them on chain.
* Making a poll for users about a certain topic.

So whatever organization we choose, it should be easy to query all posts under a topic.

The blog organization was then a simple way potentially organize all posts by a user, if they choose.

TBH I am not super attached to this structure, but i think any other structures should have at least these properties.

**Comment by gnunicorn** [comment]
Is this still being worked towards?

**Pull request closed without merging**